Source/Particles/Deposition/ChargeDeposition.H

```diff
@@ -201,6 +201,316 @@ void doChargeDepositionShapeN (const GetParticlePosition& GetPosition,
+                                     const amrex::Dim3 lo,
+                                     const amrex::Real q,
+                                     const int n_rz_azimuthal_modes,
+                                     amrex::Real* cost,
+                                     const long load_balance_costs_update_algo,
+                                     const amrex::DenseBins<WarpXParticleContainer::ParticleType>& a_bins,
+                                     const amrex::Box& box,
+                                     const amrex::Geometry& geom,
+                                     const amrex::IntVect& a_tbox_max_size)
+{
+    using namespace amrex;
+
+#if !defined(AMREX_USE_GPU)
+    amrex::ignore_unused(ix_type, cost, load_balance_costs_update_algo, a_bins, box, geom, a_tbox_max_size);
+#endif
+
+    // Whether ion_lev is a null pointer (do_ionization=0) or a real pointer
+    // (do_ionization=1)
+    const bool do_ionization = ion_lev;
+    const amrex::Real dzi = 1.0_rt/dx[2];
+#if defined(WARPX_DIM_1D_Z)
+    const amrex::Real invvol = dzi;
+#endif
+#if defined(WARPX_DIM_XZ) || defined(WARPX_DIM_RZ)
+    const amrex::Real dxi = 1.0_rt/dx[0];
+    const amrex::Real invvol = dxi*dzi;
+#elif defined(WARPX_DIM_3D)
+    const amrex::Real dxi = 1.0_rt/dx[0];
+    const amrex::Real dyi = 1.0_rt/dx[1];
+    const amrex::Real invvol = dxi*dyi*dzi;
+#endif
+
+#if (AMREX_SPACEDIM >= 2)
+    const amrex::Real xmin = xyzmin[0];
+#endif
+#if defined(WARPX_DIM_3D)
+    const amrex::Real ymin = xyzmin[1];
+#endif
+    const amrex::Real zmin = xyzmin[2];
+
+    amrex::Array4<amrex::Real> const& rho_arr = rho_fab.array();
+    auto rho_box = rho_fab.box();
+    amrex::IntVect const rho_type = rho_box.type();
+
+    constexpr int NODE = amrex::IndexType::NODE;
+    constexpr int CELL = amrex::IndexType::CELL;
+
+    // Loop over particles and deposit into rho_fab
+#if defined(WARPX_USE_GPUCLOCK)
+    amrex::Real* cost_real = nullptr;
+    if( load_balance_costs_update_algo == LoadBalanceCostsUpdateAlgo::GpuClock) {
+        cost_real = (amrex::Real *) amrex::The_Managed_Arena()->alloc(sizeof(amrex::Real));
+        *cost_real = 0.;
+    }
+#endif
+
+#if defined(AMREX_USE_GPU)
+    const auto dxiarr = geom.InvCellSizeArray();
+    const auto plo = geom.ProbLoArray();
+    const auto domain = geom.Domain();
+
+    const amrex::Box sample_tbox(IntVect(AMREX_D_DECL(0, 0, 0)), a_tbox_max_size - 1);
+    amrex::Box sample_tbox_x = convert(sample_tbox, ix_type);
+
+    sample_tbox_x.grow(depos_order);
+
+    const auto npts = sample_tbox_x.numPts();
+
+    const int nblocks = a_bins.numBins();
+    const auto offsets_ptr = a_bins.offsetsPtr();
+    const int threads_per_block = 256;
+
+    std::size_t shared_mem_bytes = npts*sizeof(amrex::Real);
+    amrex::IntVect bin_size = WarpX::sort_bin_size;
+
+    const std::size_t max_shared_mem_bytes = amrex::Gpu::Device::sharedMemPerBlock();
+
+    if (shared_mem_bytes > max_shared_mem_bytes) {
```

Comment:
This could actually become `AMREX_ALWAYS_ASSERT_WITH_MESSAGE`, if that supports value interpolation.

Reply:
Yes could you change this to `AMREX_ALWAYS_ASSERT_WITH_MESSAGE`? That also won't print on every rank if the assertion fails.